test/utils/types/Time.test.js

```diff
@@ -0,0 +1,150 @@
+require('@openzeppelin/test-helpers');
+
+const { expect } = require('chai');
+const { clock } = require('../../helpers/time');
+
+const Time = artifacts.require('$Time');
+
+const unpackDelay = delay => ({
+  valueBefore: (BigInt(delay) >> 32n) % (1n << 32n),
+  valueAfter: (BigInt(delay) >> 0n) % (1n << 32n),
+  effect: (BigInt(delay) >> 64n) % (1n << 48n),
+});
+
+const packDelay = ({ valueBefore, valueAfter = 0n, effect = 0n }) =>
+  (BigInt(valueAfter) % (1n << 32n) << 0n) +
+  (BigInt(valueBefore) % (1n << 32n) << 32n) +
+  (BigInt(effect) % (1n << 48n) << 64n);
+
+const max = (first, ...values) => values.reduce((x, y) => (x > y ? x : y), first);
+
+const MAX_UINT32 = 1n << (32n - 1n);
+const MAX_UINT48 = 1n << (48n - 1n);
+
+contract('Time', function () {
+  beforeEach(async function () {
+    this.mock = await Time.new();
+  });
+
+  describe('clocks', function () {
+    it('timestamp', async function () {
+      expect(await this.mock.$timestamp()).to.be.bignumber.equal(web3.utils.toBN(await clock.timestamp()));
+    });
+
+    it('block number', async function () {
+      expect(await this.mock.$blockNumber()).to.be.bignumber.equal(web3.utils.toBN(await clock.blocknumber()));
+    });
```

Comment:
Are we using the `blockNumber` function anywhere? I can't find it. If not it might be a better idea to remove it from the Solidity code?

Reply:
We are not using it. My idea was that this helper was providing a uint48 (timepoint format) version of blocknumber. We could use it in `Votes.sol`'s clock.

Reply:
Since we are not using it currently I would vote to remove it, it can be added later.

Reply:
Ok I see it's being used now.

Reply:
Are you ok with how its being used ?

Reply:
Yes.

My only issue with the block number in this library now is that the functions that aren't `*At` in the library automatically work with timestamps. A potential issue that I imagine is if someone uses `Delay` with block number based times, but uses `get` instead of `getAt` somewhere.

Reply:
Should we add a additional argument to the `get` that specify which mode they are operating on.

Maybe we could have a `Time.clock(<this enum>) returns (uint48)` function.

Reply:
At that point it feels overengineered really.

Reply:
so do we keep it like this ?